# EDA - WEEK 9 IP - SPAM E-mail Database

## 1) Defining the Question

### a) Specifying the Question

>Determine whether a given email is spam or not.

### b) Defining the Metric for Success

### c) Understanding the context

### d) Recording the Experimental Design

### e) Data Relevance

## 2) Reading the Data

In [1]:
# load our libraries

# data analysis
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings 
warnings.filterwarnings('ignore')

# Modelling
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.naive_bayes import MultinomialNB

print(f"Libraries imported")

Libraries imported


In [ ]:
# load our dataset
df = pd.read_csv('spambase.csv')

## 3) Checking the data

In [ ]:
# determining the no of records
df.shape

(4600, 58)

Our dataset has 58 columns and 4600 rows

In [ ]:
# previewing the top of our dataset
df.head()

0  0.64  0.64.1  0.1  0.32   0.2  ...   0.43   0.44  3.756   61   278  1
0  0.21  0.28    0.50  0.0  0.14  0.28  ...  0.180  0.048  5.114  101  1028  1
1  0.06  0.00    0.71  0.0  1.23  0.19  ...  0.184  0.010  9.821  485  2259  1
2  0.00  0.00    0.00  0.0  0.63  0.00  ...  0.000  0.000  3.537   40   191  1
3  0.00  0.00    0.00  0.0  0.63  0.00  ...  0.000  0.000  3.537   40   191  1
4  0.00  0.00    0.00  0.0  1.85  0.00  ...  0.000  0.000  3.000   15    54  1

[5 rows x 58 columns]

In [ ]:
# previewing the bottom of our dataset
df.tail()

0  0.64  0.64.1  0.1  0.32   0.2  ...  0.43  0.44  3.756  61  278  1
4595  0.31   0.0    0.62  0.0  0.00  0.31  ...   0.0   0.0  1.142   3   88  0
4596  0.00   0.0    0.00  0.0  0.00  0.00  ...   0.0   0.0  1.555   4   14  0
4597  0.30   0.0    0.30  0.0  0.00  0.00  ...   0.0   0.0  1.404   6  118  0
4598  0.96   0.0    0.00  0.0  0.32  0.00  ...   0.0   0.0  1.147   5   78  0
4599  0.00   0.0    0.65  0.0  0.00  0.00  ...   0.0   0.0  1.250   5   40  0

[5 rows x 58 columns]

In [ ]:
# check the datatypes of our dataset
df.dtypes

0         float64
0.64      float64
0.64.1    float64
0.1       float64
0.32      float64
0.2       float64
0.3       float64
0.4       float64
0.5       float64
0.6       float64
0.7       float64
0.64.2    float64
0.8       float64
0.9       float64
0.10      float64
0.32.1    float64
0.11      float64
1.29      float64
1.93      float64
0.12      float64
0.96      float64
0.13      float64
0.14      float64
0.15      float64
0.16      float64
0.17      float64
0.18      float64
0.19      float64
0.20      float64
0.21      float64
0.22      float64
0.23      float64
0.24      float64
0.25      float64
0.26      float64
0.27      float64
0.28      float64
0.29      float64
0.30      float64
0.31      float64
0.32.2    float64
0.33      float64
0.34      float64
0.35      float64
0.36      float64
0.37      float64
0.38      float64
0.39      float64
0.40      float64
0.41      float64
0.42      float64
0.778     float64
0.43      float64
0.44      float64
3.756     float64
61        

In [ ]:
# check percentage of nan values in the dataset
df.isnull().mean()*100

0         0.0
0.64      0.0
0.64.1    0.0
0.1       0.0
0.32      0.0
0.2       0.0
0.3       0.0
0.4       0.0
0.5       0.0
0.6       0.0
0.7       0.0
0.64.2    0.0
0.8       0.0
0.9       0.0
0.10      0.0
0.32.1    0.0
0.11      0.0
1.29      0.0
1.93      0.0
0.12      0.0
0.96      0.0
0.13      0.0
0.14      0.0
0.15      0.0
0.16      0.0
0.17      0.0
0.18      0.0
0.19      0.0
0.20      0.0
0.21      0.0
0.22      0.0
0.23      0.0
0.24      0.0
0.25      0.0
0.26      0.0
0.27      0.0
0.28      0.0
0.29      0.0
0.30      0.0
0.31      0.0
0.32.2    0.0
0.33      0.0
0.34      0.0
0.35      0.0
0.36      0.0
0.37      0.0
0.38      0.0
0.39      0.0
0.40      0.0
0.41      0.0
0.42      0.0
0.778     0.0
0.43      0.0
0.44      0.0
3.756     0.0
61        0.0
278       0.0
1         0.0
dtype: float64

## 4) Implementing the Solution

In [ ]:
# Formulate our data to X and y
X = df.drop('1', axis=1)
y = df['1']

### Naive Bayes Classifier

In [ ]:
# splitting the dataset into Training Set and Test set

# partition dataset to 80-20
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [ ]:
# Fitting our model 
 
model = MultinomialNB().fit(X_train, y_train)

In [ ]:
# Evaluating the Model
# 

predicted = model.predict(X_test)
print(confusion_matrix(y_test, predicted))
print('*'*100)
print()
print(classification_report(y_test, predicted))
print('*'*100)
print()
print(np.mean(predicted == y_test))
print('*'*100)
print()

[[450  88]
 [ 87 295]]
****************************************************************************************************

              precision    recall  f1-score   support

           0       0.84      0.84      0.84       538
           1       0.77      0.77      0.77       382

    accuracy                           0.81       920
   macro avg       0.80      0.80      0.80       920
weighted avg       0.81      0.81      0.81       920

****************************************************************************************************

0.8097826086956522
****************************************************************************************************



*From our results our Multinomial Naive Bayes Classifier model obtained an accuracy of 81% from a 80-20 split*

*The model managed to get 945 correct predictions and 175 incorrect predictions*

## 5) Challenging the Solution

**Challenging Solution with 70-30 split**

In [ ]:
# partition dataset to 70-30
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0)

In [ ]:
# Fitting our model 
model = MultinomialNB().fit(X_train, y_train)

In [ ]:
# Evaluating the Model
# 
predicted = model.predict(X_test)
print(confusion_matrix(y_test, predicted))
print('*'*100)
print()
print(classification_report(y_test, predicted))
print('*'*100)
print()
print(np.mean(predicted == y_test))
print('*'*100)
print()

[[684 138]
 [131 427]]
****************************************************************************************************

              precision    recall  f1-score   support

           0       0.84      0.83      0.84       822
           1       0.76      0.77      0.76       558

    accuracy                           0.81      1380
   macro avg       0.80      0.80      0.80      1380
weighted avg       0.81      0.81      0.81      1380

****************************************************************************************************

0.8050724637681159
****************************************************************************************************



*From our results our Multinomial Naive Bayes Classifier model obtained an accuracy of 81% from a 70-30 split*

*The model managed to get 1,111 correct predictions and 269 incorrect predictions*

**Challenging Solution with a 60-40 split**

In [ ]:
# partition dataset to 60-40
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4, random_state = 0)

In [ ]:
# Fitting our model 
model = MultinomialNB().fit(X_train, y_train)

In [ ]:
# Evaluating the Model
# 
predicted = model.predict(X_test)
print(confusion_matrix(y_test, predicted))
print('*'*100)
print()
print(classification_report(y_test, predicted))
print('*'*100)
print()
print(np.mean(predicted == y_test))
print('*'*100)
print()

[[913 184]
 [198 545]]
****************************************************************************************************

              precision    recall  f1-score   support

           0       0.82      0.83      0.83      1097
           1       0.75      0.73      0.74       743

    accuracy                           0.79      1840
   macro avg       0.78      0.78      0.78      1840
weighted avg       0.79      0.79      0.79      1840

****************************************************************************************************

0.7923913043478261
****************************************************************************************************



*From our results our Multinomial Naive Bayes Classifier model obtained an accuracy of 79% from a 60-40 split*

*The model managed to get 1,458 correct predictions and 382 incorrect predictions*

## **CONCLUSION**

*From our analysis our Multinomial Naive Bayes Classifier obtained an accuracy of 81% from the 80-20 split, an accuray of 81% from the 70-30 split and an accuracy of 79% from the 60-40 split*

*From our findings it is evident that the accuracy went on decreasing as the splits continued*

*Finally the model was able to get 945 correct predictions from the 80-20 split, 1,111 correct predictions from the 70-30 split and finally 1,458 correct predictions from the 60-40 split.*

## **RECOMMENDATION**

*From our analysis a 60-40 split would be effective as the number of correct predictions from this split was high at 1,458. But based on our accuracy a 80-20 split had more accuracy at 81%*